### PR TITLE
[downsampler] Avoid mutating global DropPipelineMetadatas

### PR DESCRIPTION
**Pull request by wesleyk**
**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

**Review by arnikola — APPROVED** [review]
LGTM, nice catch!

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go [hunk 1]
```diff
@@ -394,7 +394,8 @@ func (a *metricsAppender) SamplesAppender(opts SampleAppenderOptions) (SamplesAp
 	// Apply drop policies results
 	a.curr.Pipelines, dropApplyResult = a.curr.Pipelines.ApplyOrRemoveDropPolicies()
 
-	if len(a.curr.Pipelines) > 0 && !a.curr.IsDropPolicyApplied() {
+	// Skip sending to downsampler if there's a drop policy or no pipeline defined.
+	if len(a.curr.Pipelines) > 0 && !a.curr.IsDropPolicyApplied() && !a.curr.IsDefault() {
```
Comment (ryanhall07) [inline src/cmd/services/m3coordinator/downsample/metrics_appender.go:398]: shouldn't `len(a.curr.Pipelines) > 0` already prevent sending the default staged metadata?

File: src/metrics/metadata/metadata.go [hunk 2]
```diff
@@ -273,13 +273,18 @@ func (metadatas PipelineMetadatas) ApplyOrRemoveDropPolicies() (
...
+	result := metadatas
+
+	// Drop is effective as no other non drop pipelines, result is a drop
 	if nonDropPipelines == 0 {
-		// Drop is effective as no other non drop pipelines, result is a drop
-		return DropPipelineMetadatas, AppliedEffectiveDropPolicyResult
+		// nb: Do not directly return DropPipelineMetadatas, as the client could potentially save that reference
+		// and modify global state.
+		result = result[:0]
+		result = append(result, DropPipelineMetadata)
```
Comment (ryanhall07) [inline src/metrics/metadata/metadata.go:283]: seems bad to return `AppliedEffectiveDropPolicyResult` as well, somebody could in theory alter that too (even though it's not happening today)

Comment (ryanhall07) [reply src/metrics/metadata/metadata.go:283]: i think we should refactor all of these global vars into accessor methods that return copies of package protected vars.

**Review by ryanhall07 — APPROVED** [review]
LG. I think there is some follow up work here to prevent future regressions.

File: src/cmd/services/m3coordinator/downsample/metrics_appender.go (hunk 1 above)
Comment (wesleyk) [reply src/cmd/services/m3coordinator/downsample/metrics_appender.go:398]: the default confusingly contains one PipelineMetadata (so len == 1) which in itself is "empty" / default.

**Pull request merged**